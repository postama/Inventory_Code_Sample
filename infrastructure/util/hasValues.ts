export function hasValues(object: object): boolean {
    return object && Object.keys(object).length > 0;
}

export function isEmpty(object: object): boolean {
    return !hasValues(object);
}