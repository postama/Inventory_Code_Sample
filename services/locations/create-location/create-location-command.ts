import { literal, intersection, type, TypeOf } from 'io-ts';
import { BaseLocationCommand } from '../shared/base-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import uuid from 'uuid/v4';

export const CreateLocationCommandValidator = intersection([
    BaseLocationCommand,
    type({ type: literal('CreateLocation') })
]);

export type CreateLocationCommand = TypeOf<typeof CreateLocationCommandValidator>;

export function createCreateLocationCommand(request: any): CreateLocationCommand {
    return validateType({ ...request, isActive: true, type: 'CreateLocation', locationId: uuid() }, CreateLocationCommandValidator);
}