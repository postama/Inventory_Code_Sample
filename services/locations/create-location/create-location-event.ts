import { intersection, type, literal, TypeOf } from 'io-ts';
import { createLocation, LocationValidator } from '../../../repo/locations/location-type';
import { CreateLocationCommand } from './create-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';
import uuid from 'uuid/v4';

export const CREATE_LOCATION = 'LocationCreated';

export const LocationCreatedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(CREATE_LOCATION),
        version: literal(1),
        data: LocationValidator
    })
]);

export type LocationCreatedV1 = TypeOf<typeof LocationCreatedValidatorV1>;

export function createCreateLocationEvent(command: CreateLocationCommand): LocationCreatedV1 {
    const event = {
        ...createBaseEvent(command, uuid(), CREATE_LOCATION, 1),
        data: createLocation(command)
    };

    return validateType(event, LocationCreatedValidatorV1);
}