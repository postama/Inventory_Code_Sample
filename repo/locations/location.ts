import * as dynamo from '../../infrastructure/persistence/dynamo';
import { Reducer } from '../../infrastructure/reducer/reducer';
import { Location } from './location-type';
import { LocationEvent } from './location-event-types';
import { CREATE_LOCATION } from '../../services/locations/create-location/create-location-event';
import { createLocationReducer } from '../../services/locations/create-location/create-location-reducer';
import { getEntityQuery } from '../events/entity';
import { EDIT_LOCATION } from '../../services/locations/edit-location/edit-location-event';
import { editLocationReducer } from '../../services/locations/edit-location/edit-location-reducer';

export const LocationReducer = new Reducer<Location, LocationEvent>();

LocationReducer.register(CREATE_LOCATION, createLocationReducer);
LocationReducer.register(EDIT_LOCATION, editLocationReducer);

export async function getById(customerId: string, locationId: string): Promise<Location> {
    return await dynamo.pagedReducer<Location, LocationEvent>(LocationReducer, getEntityQuery(customerId, locationId));
}

export async function save(event: LocationEvent) {
    return await dynamo.put(event);
}