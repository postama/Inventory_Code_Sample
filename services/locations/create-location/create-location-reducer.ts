import { Location } from '../../../repo/locations/location-type';
import { LocationCreatedV1 } from './create-location-event';

// tslint:disable-next-line variable-name
export function createLocationReducer(_location: Location, event: LocationCreatedV1): Location {
    return {
        ...event.data,
        siteId: event.siteId,
        created: event.created
    };
}