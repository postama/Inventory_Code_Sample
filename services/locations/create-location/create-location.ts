import { APIGatewayEvent } from 'aws-lambda';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { CreateLocationCommand, createCreateLocationCommand } from './create-location-command';
import { checkUniqueCommand } from '../../../repo/events/event';
import { LocationCreatedV1, createCreateLocationEvent } from './create-location-event';
import { createLocationReducer } from './create-location-reducer';
import { save } from '../../../repo/locations/location';

export function entry(request: APIGatewayEvent) {
    return requestHandler(createLocation, request);
}

export async function createLocation(request): Promise<void> {
    const command: CreateLocationCommand = createCreateLocationCommand(request);
    await checkUniqueCommand(command);
    const event: LocationCreatedV1 = createCreateLocationEvent(command);
    event.state = createLocationReducer({} as any, event);
    await save(event);
    return;
}