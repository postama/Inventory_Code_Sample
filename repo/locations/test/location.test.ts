import { createTestCreateLocationEvent } from '../../../services/locations/create-location/tests/test.util';
import { save, getById } from '../location';
import { LocationCreatedV1 } from '../../../services/locations/create-location/create-location-event';

describe('location aggregate root', () => {
    describe('getById', () => {
        test('should get a location', async () => {
            const event: LocationCreatedV1 = createTestCreateLocationEvent();
            await save(event);
            const location = await getById(event.customerId, event.entityId);
            expect(location).toMatchSnapshot({
                created: expect.any(String)
            });
        });
    });
});