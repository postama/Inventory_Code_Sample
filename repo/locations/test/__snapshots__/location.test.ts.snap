// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`location aggregate root getById should get a location 1`] = `
Object {
  "created": Any<String>,
  "isActive": true,
  "name": "abc",
  "siteId": "TESTER",
}
`;
